[WIP] Allow getting factories from component Host

tigrannajaryan opened the pull request:
Added GetFactory function to Host interface. This allows components
to fetch factories and create components dynamically. The functionality
will be used by receiver creator that creates and destroys other receivers
as it discovers relevant external endpoints from which these receivers
can get data.

-------

This PR shows 2 different approaches.

The first commit in this PR shows the "minimal change" approach to add the functionality we need: https://github.com/tigrannajaryan/opentelemetry-service/pull/3/commits/967e2ff18136f43a30e2bbba57996d22aa015a69
The slight downside is that `Host.GetFactory` does not return the most narrow type possible for the component kind.

The second commit introduces 4 narrower-typed `Host.GetFactory` functions. However this requires more refactoring and introduces slightly weird set of factory types. We have the following composition for factories:

component.Factory -> component.ReceiverFactory -> receiver.BaseFactory -> receiver.Factory

I am not totally convinced that the second commit is an improvement.

Commit:
Allow getting factories from component Host

Added GetFactory function to Host interface. This allows components
to fetch factories and create components dynamically. The functionality
will be used by receiver creator that creates and destroys other receivers
as it discovers relevant external endpoints from which these receivers
can get data.

jrcamp commented:
I don't have a preference either way. I'd be fine with the simpler one. I'll throw out a third alternative that would require reflection, not sure it's any better:

```go
var factory factory.ReceiverFactory
if err := GetFactory(type, &factory.ReceiverFactory); err != nil {
   return err
}
// use receiver factory
```

tigrannajaryan commented:
> I don't have a preference either way. I'd be fine with the simpler one. I'll throw out a third alternative that would require reflection, not sure it's any better:
> 
> ```go
> var factory factory.ReceiverFactory
> if err := GetFactory(type, &factory.ReceiverFactory); err != nil {
>    return err
> }
> // use receiver factory
> ```

Thanks for suggesting an alternate, I didn't think of this. This eliminates the need to pass the `kind`, but requires returning an err since there is no other way to indicate factory was not found. Also, I think it requires GetFactory to declare its second parameter as `interface{}`, which means we get even less support from the compiler to catch mistakes.

If we see other benefits we can go this way otherwise let's just stick with the first approach.

bogdandrutu commented on component/component.go:
```diff
@@ -40,4 +51,23 @@ type Host interface {
+	// until Shutdown() is called. Note that the component is responsible for destroying
+	// other components that it creates.
+	GetFactory(kind Kind, componentType string) Factory
+
```
Do we need this generic method?

tigrannajaryan replied:
This is showing 2 different approach. We either need the generic one or the 4 others.

The 4 methods are the approach that you asked for, but I am not sure it is a good idea. It requires more changes and gains are unclear. The composition of interfaces looks confusing to me.

See commit 1 vs commit 1+2 in this PR.

owais commented:
Second one doesn't seem to be a great improvement to me either. There could also be helpers that can reduce any type checking boilerplate with the first one. I also like @jrcamp's suggestion to read into a typed pointer.

pjanotti committed:
Add Monitoring section and doc (#682)

* Add Monitoring section and doc

Add a quick blurb about monitoring to the README.md and a monitoring.md
with some key recommendations to users.

* Fix typo

* Remove changes to README.md

This conflicts with a re-org on README.md.

bogdandrutu committed:
Change Metric, MetricDescriptor and use the new slices for Metric (#678)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Automatically generate tests (#679)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

flands committed:
Beta README.md (#581)

Beta README.md

pjanotti committed:
Improve OC metrics observability and return errors to clients (#659)

* Improve OC metrics observability and return errors to clients

* Make OC Trace to use value semantics internally

* Make OC metrics and trace return err if next consumer returns error

* Avoid logging the same error repeatdly on testbed

bogdandrutu committed:
Address late feedback from #660 by fixing generated comments

tigrannajaryan committed:
Enable OTLP exporter and add OTLP E2E tests (#684)

- OTLP exporter is now a component enabled in the default build.
- Added OTLP E2E tests for traces.

TODO: Add OTLP E2E tests for metrics once metrics are supported
in OTLP exporter.

Resolves: https://github.com/open-telemetry/opentelemetry-collector/issues/668

dmitryax committed:
Add OC -> OTLP metrics translation (#687)

- Used in OTLP Exporter to translate internal OC format to OTLP
- translator/common/oc_to_otlp.go code was moved as is from trace/oc_to_otlp.go

Testing: Unit tests

bogdandrutu committed:
Update proto deps to v0.3.0 (#686)

As part of this replace all Unixnano -> UnixNano

bogdandrutu committed:
Fix build after #686 was merged (#693)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Automatically generate Span/SpanLink and slices for events and links (#680)

* Automatically generate Span/SpanLink and slices for events and links

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

* Use structs instead of origins in tests

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Enable static-check and add gosec (not enabled) (#677)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Remove otlp_to_oc, redundant with MetricDataFromOtlp (~0 overhead) + MetricDataToOC (#685)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

dmitryax committed:
Add e2e test for OLTP metrics and fix a bug in oc->otlp translation (#694)

- Add an e2e test for OLTP metrics
- Fix a bug in OC->OTLP translation when metrics with no labels were failing

dmitryax committed:
Fix the build broken due to auto-rebase (#696)

tigrannajaryan committed:
Refactor component and factory interface definitions (#683)

As we are preparing for Beta release we want to cleanup publicly exported
types and interfaces and do all necessary refactoring and breaking changes
now, before the Beta release. We will have a lot less leeway for breaking
changes after the Beta release.

Component related type declarations are now all in the `component` package.
This makes it possible for the interfaces to reference each other. This
was were very restricted earlier because component interfaces were in 5
different packages and many proposals were impossible to implement because
they would result in circular dependencies between packages.

(An example upcoming new capability that is enabled by this refactoring
is for components to query the Host for other components and for factories).

List of changes in this commit:

- Move all factory interfaces and component interfaces to component package.
- Rename old factories and components interfaces to use "Old" suffix for clarity.
- Eliminate forced checks that components implement factories. This is already
  enforced by the compiler when the factory is added to the Defaults() and
  was unnecessary code.
- Eliminated some unnecessary codes (removed overall over 200 lines).
- Run `go mod tidy` on testbed.

Warning: this is a breaking change to publicly exported types and function
signatures. We announced that a breaking change is comming. Once we agree
to merge this commit we will need to announce the exact list of changes
and guide component authors to modify their components accordingly.

Future changes:
- Once all components are migrated to the new internal representation,
  delete all "Old"-suffixed definitions. This will likely be done while
  we are still in Beta phase, before the Stable release.

bogdandrutu committed:
Improve tests and fix small issues in  metrics_to_oc and internal (#692)

toumorokoshi committed:
Make prometheus reciever config loading strict (#697)

Fixes #583

Using strict umarshalling ensures eliminates common mistakes
such as adding a key intended for the prometheus reciever itself
into the prometheus subconfiguration.

bogdandrutu committed:
Update more V2 code to Old and non-v2 (#700)

* Update more V2 code to Old and non-v2

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

* Update component/receiver.go

Co-Authored-By: Tigran Najaryan <4194920+tigrannajaryan@users.noreply.github.com>

Co-authored-by: Tigran Najaryan <4194920+tigrannajaryan@users.noreply.github.com>

bogdandrutu committed:
Generate all trace structs automatically. (#695)

* Generate all trace structs automatically.

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

* Fix review comments

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

lizthegrey committed:
cross-compile linux/arm64, darwin/amd64, windows/amd64 (#701)

lizthegrey committed:
re-enable go vet, and fix warnings. (#703)

* re-enable go vet, and fix warnings.

* manually hack versions around.

dmitryax committed:
Add internal data -> OC compatibility shims to processors and exporters (#667)

In order to start migrating components from OC to internal data, all parts of the pipeline should be able to automatically convert back to OC if downstream component doesn't work with internal data structure

bogdandrutu committed:
Fix G07 errors in gosec evaluation (#704)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Fix G302 errors in gosec evaluation (#705)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Fix G304 errors in gosec evaluation (#706)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Upgrade to the latest viper release (#707)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

lizthegrey committed:
initial switch to circleci (#702)

* initial switch to circleci

* update docs

* test packages in batches of 10.

* update golangci to have go1.14 vet support

* roll back go.mod, move to separate pull.

* allow running longer to avoid circleci timeout.

* further increase timeout.

* further increase timeout.

* fix data race, adjust timeout.

* 10 min timeout (will decrease with parallelism later)

* batch/parallelize all test executions.

* modularize circleci

* add loadtest tools

* optimize running time

* address review feedback

lizthegrey committed:
Fix CircleCI errors: ignore if cache missing (#715)

Also fixes flaky bench tests

bogdandrutu committed:
Add to/from OTLP for trace (#714)

* Add to/from OTLP for trace

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

* Fix typo

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

lizthegrey committed:
parallelize loadtests & cross-compiles (#716)

* parallelize loadtests.

* print tests to be run by each worker.

* listtests -s

* eliminate excess quote

* build more in parallel.

* parallelize lint, in-container

* fix broken variables.

* use more robust parallelism

bogdandrutu committed:
Remove otlp to/from oc converters and use internal converters (#720)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

dmitryax committed:
Add internal data support to cloning fanout connectors (#717)

Cloning fanout connectors now able to work with new type of downstream consumers.  

- Also add test data generators for metric internal data from https://github.com/open-telemetry/opentelemetry-collector/pull/718

bogdandrutu committed:
Report more accurate code coverage (#723)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

owais committed:
Publish to otel dockerhub account (#721)

- Publish stable builds from release tags to
otel/opentelemetry-collector-contrib
- Publish dev builds from master commits to
otel/opentelemetry-collector-contrib-dev

bogdandrutu committed:
Remove cleverness that defined tests cases (#730)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

pjanotti committed:
Changes after `go mod tidy` on testbed (#733)

pjanotti committed:
Add service.instance.id to metrics generated by the Collector (#732)

* Add service.instance.id to metrics generated by the Collector

This allows each instance of the collector to be uniquely identified in any deployment.

* Add an option to control addition of instance ID label

* gofmt -s -w service/telemetry.go

bogdandrutu committed:
Remove unnecessary .nocover, there is an empty_test.go file there. (#727)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Change publish jobs to attach to the correct workspace (#731)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Enable more govet checks, fix shadows (#734)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Enable all govet analyzers. (#739)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Enable gosec in CI, takes ~50sec so make it separate job. (#738)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

Commit:
Enable best effort serve by default of Prometheus Exporter (#713)

Signed-off-by: Liam White <liam@tetrate.io>

bogdandrutu committed:
Use linters and config from the core for testbed (#740)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Fix null pointer exception in the logging exporter (#743)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Remove unnecessary condition to have at least one processor (#744)

This makes testing the collector locally harder. Also the condition is not enforced because having a disabled processor in the pipeline passes that condition which makes it not very useful.

owais committed:
Automate publishing of Github releases (#745)

Now we'll automatically create Github releases from tags and upload the
binaries and their checksums as release artifacts.

owais committed:
Disbaled publish-dev job temporarily (#746)

publish-dev job needs to be in it's own workflow as it ignores all tags
and that conflicts with the publish-stable workflow. I'll re-enable
publish-dev in a separate PR later.

pjanotti committed:
Fix race in TestExportProtocolViolations_nodelessFirstMessage (#747)

Fix race in TestExportProtocolViolations_nodelessFirstMessage. The race is on the t *testing.T having one of its fields being read and write concurrently. I didn't investigate why this test needs this special timeout handling, just took the obvious fix to protect against the data race - will look at this again later.

Fixes https://github.com/open-telemetry/opentelemetry-collector/issues/41

owais committed:
Enable publishing releases from tags (#749)

Circle was not running any builds for tags because the publish job had
other jobs as dependencies and they did not explicitly specify that they
should be run for tags. Looks like Circle will always run a job for any
branch unless the branch is ignored and never run a job for a tag unless
the tag is explicitly enabled.

Co-authored-by: Bogdan Drutu <bogdandrutu@gmail.com>

owais committed:
Re-enable publishing dev docker images from master (#753)

This PR enables publishing dev docker images from every commit to
master. This was disabled earlier as I suspected it was causing issues
with the job that publishes images from tags.

bogdandrutu committed:
Remove set functions for non-primitives and add a Has function (#691)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

objectiser committed:
[BREAKING CHANGE] Remove deprecated Jaeger tchannel receiver (#636)

* Remove deprecated Jaeger tchannel receiver

* Fix minor issues

* Rebase

* Fix test

objectiser committed:
[BREAKING CHANGE] Remove the jaeger thrift exporter, and make grpc the single exporter for Jaeger (#629)

* Remove Jaeger thrift exporter and make the gRPC exporter the one and only Jaeger exporter

Signed-off-by: Gary Brown <gary@brownuk.com>

* Fix the testbed

Signed-off-by: Gary Brown <gary@brownuk.com>

* Rebase

bogdandrutu committed:
Remove unused variable. (#755)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Remove unused translator oc to jaegerthrift (#757)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

pjanotti committed:
Update examples to latest published dev image (#761)

These examples currently are linked in the docs, the docs should have examples against stable releases.

bogdandrutu committed:
Move Init function from parent to the struct (#758)

* Move Init function from parent to the struct

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

* Fix review comments

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

tigrannajaryan committed:
Restore Vision document (#764)

This was lost accidentally in https://github.com/open-telemetry/opentelemetry-collector/pull/581

bogdandrutu committed:
Prepare AttributeKeyValue to allow removing AttributeValue (#759)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>
